File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def parse_time(time):
    hour, minute = time.split(":")
    return int(hour)*60 + int(minute)


def get_playtime(start_time, end_time):
    return parse_time(end_time) - parse_time(start_time)


def parse_music(music: str):
    return music.replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g").replace("A#", "a")


def extend_music(akbo, playtime):
    a, b = divmod(playtime, len(akbo))
    return a * akbo + akbo[:b]


def solution(melody, musicinfos):
    answer = []
    for music in musicinfos:
        melody = parse_music(melody)
        start_time, end_time, music_name, akbo = music.split(",")
        playtime = get_playtime(start_time, end_time)
        akbo = parse_music(akbo)
        akbo = extend_music(akbo, playtime)

        if melody in akbo:
            answer.append([music_name, playtime])

    answer.sort(key=lambda x: -x[1])
    return answer[0][0] if answer else "(None)"

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public String solution(String m, String[] musicinfos) {
        List<String> answer = new ArrayList<>();
        List<Integer> answerTime = new ArrayList<>();

        m = replaceAcbo(m);

        for (int i = 0; i < musicinfos.length; i++) {
            String[] split = musicinfos[i].split(",");

            String from = split[0];
            String to = split[1];
            int time = getTime(from, to);

            String title = split[2];
            String melody = extendMelody(replaceAcbo(split[3]), time);

            if (melody.contains(m)) {
                answer.add(title);
                answerTime.add(time);
            }
        }

        if (answer.isEmpty()) {
            return "(None)";
        }

        int max = 0;
        int maxIndex = 0;

        for (int i = 0; i < answerTime.size(); i++) {
            if (answerTime.get(i) > max) {
                max = answerTime.get(i);
                maxIndex = i;
            }
        }

        return answer.get(maxIndex);
    }

    public String replaceAcbo(String from) {
        from = from.replace("C#", "c");
        from = from.replace("B#", "b");
        from = from.replace("D#", "d");
        from = from.replace("F#", "f");
        from = from.replace("G#", "g");
        from = from.replace("A#", "a");

        return from;
    }

    public int getTime(String from, String to) {
        String[] fromArr = from.split(":");
        String[] toArr = to.split(":");

        int fromHour = Integer.parseInt(fromArr[0]);
        int fromMinute = Integer.parseInt(fromArr[1]);

        int toHour = Integer.parseInt(toArr[0]);
        int toMinute = Integer.parseInt(toArr[1]);

        return (toHour - fromHour) * 60 + (toMinute - fromMinute);
    }

    public String extendMelody(String melody, int time) {
        int n = time / melody.length();
        int m = time % melody.length();

        return melody.repeat(n) + melody.substring(0, m);
    }
}